Add Atmel SAMA5 platform support

nodeax opened the pull request:
Adding support for Atmel/Microchip [SAMA5D2-XULT board](http://www.atmel.com/tools/atsama5d2-xult.aspx)

The processor supports: 
ARM TrustZone, configuring memory and/or peripherals as secure, secure RTC, Secure boot, On-the-fly encryption/decryption of the DDR bus, Tamper protection etc. [Datasheet](http://www.atmel.com/Images/Atmel-11267-32-bit-Cortex-A5-Microcontroller-SAMA5D2_Datasheet.pdf).

Instructions to build optee_os along with bootlog, test results can be found [here](https://github.com/nodeax/optee_os/wiki).

MrVan commented on core/arch/arm/plat-sam/platform_config.h:
```diff
@@ -0,0 +1,134 @@
+ * +------------------+
+ * |   Linux memory   |
+ * +--------+---------+
+ */
```
So optee runs in low memory, Linux runs in high memory? To ARM32 Linux, I recall that there is limitation that linux image needs to be put in the lower 128M address space. If OP-TEE runs in low memory, you may need to reserve the lower 128M to OP-TEE and linux could not use that.

nodeax commented:
Realized the memory map in the comment needs to be updated. optee runs in
the middle of the DDR (as shown in SAMA5D2 memory map).

Long story: As you said on ARM32 linux putting optee in lower 128M means
Linux can not use that memory. However I have an additional restriction on
SAMA5D2 where memory protection works in chunks of 128MiB. Inside the
128MiB chunk the lower memory has fine resolution for configuring as secure
memory but high memory can only be configured as a single 64MiB secure
memory chunk. So to avoid unnecessarily setting aside 64MiB for secure
memory, I put op-tee at the start of a 128MiB boundary in DDR.

End of DDR memory
+ * +------------------+
+ * |   Linux memory   |
+ * +------------------+
+ * |      SHMEM       |
+ * +------------------+
+ * |        | TA_RAM  |
+ * + TZDRAM +---------+
+ * |        | TEE_RAM |
+ * +------------------+
+ * |   Linux memory   |
+ * +--------+---------+
Start of DDR memory

/*
 * SAMA5D2 memory map
 *
 *  0x4000_0000                                  -
 *    Linux/other                                |
 *  0x30C0_0000                                  -
 *    Shared memory: 4 MiB (CFG_SHMEM_SIZE)      |
 *  0x3080_0000 [CFG_SHMEM_START]                -
 *    TA RAM: 7 MiB (remaining memory)           |
 *  0x3010_0000 [CFG_TA_RAM_START]               -
 *    TEE RAM: 1 MiB (CFG_TEE_RAM_VA_SIZE)       |
 *  0x3000_0000 [CFG_TEE_RAM_START/LOAD_ADDR]    -
 *    Linux/other                                |
 *  0x2000_0000 [DRAM0_BASE]                     -
 *
 */

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,610 @@
+		.matrix = MATRIX_H64MX,
+		.security_type = SECURITY_TYPE_PS,
+	},
+	/* ARM */
```
This comment (and all the others below) are kind of useless since they simply repeat what's written into `.peri_id`, they don't really improve readability either IMO. But you can keep them if you prefer.

jforissier commented on core/arch/arm/plat-sam/platform_config.h:
```diff
@@ -0,0 +1,134 @@
+ * Treats shared accesses (bit22=0, bit13=0)
+ * Parity disabled (bit21=0)
+ * Event monitor disabled (bit20=0)
+ * Platform fmavor specific way config:
```
s/fmavor/flavor/

jforissier commented on core/arch/arm/plat-sam/platform_config.h:
```diff
@@ -0,0 +1,134 @@
+ *
+ * I/Dcache prefetch enabled (bit29:28=2b11)
+ * NS can access interrupts (bit27=1)
+ * NS can lockown cache lines (bit26=1)
```
s/lockown/lockdown/

jforissier commented on core/arch/arm/plat-sam/tz_matrix.h:
```diff
@@ -0,0 +1,113 @@
+#ifndef __TZ_MATRIX_H__
+#define __TZ_MATRIX_H__
+
+#define MATRIX_MCFG(n)	(0x0000 + (n) * 4)/* Master Configuration Register */
```
Space after `)` and before `/*`

jforissier commented on core/include/drivers/atmel_uart.h:
```diff
@@ -0,0 +1,40 @@
+};
+
+void atmel_uart_init(struct atmel_uart_data *pd, paddr_t base);
+void set_mmu_flag(void);
```
Unused? (remove)

nodeax commented:
@jforissier : Thanks for all the review comments/suggestions. All of them have been addressed in the updated pull request.
@MrVan: The memory map comment has been updated to reflect the correct layout.

igoropaniuk commented on core/arch/arm/plat-sam/tz_matrix.h:
```diff
@@ -0,0 +1,113 @@
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __TZ_MATRIX_H__
```
minor: different styles of header guards are used through all header files (sometimes wrapped with `__ __`, sometimes without them)

nodeax replied:
Thanks, fixed in latest update.

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+
+	for (i = 0; i < ARRAY_SIZE(peri_security_array); i++) {
+		if (peri_id == peri_security_array[i].peri_id)
+			break;
```
`return &peri_security_array[i];` here and `return NULL;` after the loop.

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	matrix_write(matrix_base, MATRIX_SASSR(slave), srsplit_setting);
+}
+
+static struct peri_security *get_peri_security(unsigned int peri_id)
```
Should probably return `const struct peri_security *`

jforissier commented on core/arch/arm/plat-sam/matrix.h:
```diff
@@ -0,0 +1,44 @@
+					unsigned int size);
+
+extern vaddr_t matrix32_base(void);
+extern vaddr_t matrix64_base(void);
```
Are those `extern` needed? I don't think so.

nodeax replied:
matrix32/64_base() is being used in both matrix.c and main.c, hence it is included in the header.

I have fixed all other review comments in the updated pull request. Thanks for reviewing.

jforissier replied:
Huh, sorry for not being clear, I meant the `extern` keyword is not needed, is it?

nodeax replied:
sorry, misread the original comment. no they are not needed, fixed in latest update.

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	struct peri_security *periperal_sec;
+	int ret;
+
+	if ((peri_id_array == NULL) || (size == 0))
```
`if (!peri_id_array || !size)`

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	peri_id_p = peri_id_array;
+	for (i = 0; i < size; i++) {
+		periperal_sec = get_peri_security(*peri_id_p);
+		if (periperal_sec == NULL)
```
`if (!peripheral_sec)`

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	unsigned int *peri_id_p;
+	unsigned int matrix;
+	unsigned int peri_id;
+	struct peri_security *periperal_sec;
```
Typo: `s/periperal/peripheral/`

jforissier commented on core/arch/arm/plat-sam/main.c:
```diff
@@ -0,0 +1,405 @@
+static void matrix_configure_slave_h64mx(void)
+{
+	unsigned int ddr_port;
+	unsigned int ssr_setting, sasplit_setting, srtop_setting;
```
One declaration per line, please.

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+struct peri_security {
+	unsigned int	peri_id;
+	unsigned int	matrix;
+	unsigned int	security_type;
```
No tabs between type and member name, please. Use one space instead.

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	},
+};
+
+static inline void matrix_write(unsigned int base,
```
Drop inline

jforissier commented on core/arch/arm/plat-sam/matrix.c:
```diff
@@ -0,0 +1,537 @@
+	write32(value, offset + base);
+}
+
+static inline unsigned int matrix_read(int base, unsigned int offset)
```
Drop inline

jforissier commented on core/arch/arm/plat-sam/main.c:
```diff
@@ -0,0 +1,405 @@
+	AT91C_ID_CHIPID,
+};
+
+static int matrix_config_periheral(void)
```
Typo, and perhaps plural is better? (matrix_config_peripherals)

jforissier commented on core/arch/arm/plat-sam/main.c:
```diff
@@ -0,0 +1,405 @@
+	if (ret)
+		return -1;
+
+	return 0;
```
Simplify:
```C
static int matrix_config_peripherals(void)
{
        return matrix_configure_peri_security(security_ps_peri_id,
                                              ARRAY_SIZE(security_ps_peri_id));
}
```

jforissier commented on core/arch/arm/plat-sam/main.c:
```diff
@@ -0,0 +1,405 @@
+	if (ret)
+		return -1;
+
+	return 0;
```
Simplify: `return matrix_config_peripherals();`

jenswi-linaro commented:
Is this review still active?

jforissier commented:
LGTM, @nodeax you may add my Ack to the 2nd commit, too.
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

nodeax commented:
@jforissier : Thanks for the review, added you Ack to the 2nd commit.
@jenswi-linaro : Was on vacation. Rebased/Verified everything still works after pulling in changes from master and also added Ack from jforissier.

jforissier commented:
@nodeax thanks for contributing!
